File: src/mediathek/controller/IoXmlSchreiben.java
/*
 * MediathekView
 * Copyright (C) 2008 W. Xaver
 * W.Xaver[at]googlemail.com
 * http://zdfmediathk.sourceforge.net/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package mediathek.controller;

import java.io.IOException;
import java.io.OutputStreamWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import javax.xml.stream.XMLOutputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamWriter;
import mediathek.daten.Daten;
import mediathek.daten.DatenAbo;
import mediathek.daten.DatenBlacklist;
import mediathek.daten.DatenDownload;
import mediathek.daten.DatenProg;
import mediathek.daten.DatenPset;
import mediathek.tool.Konstanten;
import mediathek.tool.MVConfig;
import mediathek.tool.MVReplaceList;
import msearch.filmeLaden.DatenFilmlisteUrl;
import msearch.tool.MSConst;

public class IoXmlSchreiben {

    private static XMLStreamWriter writer;
    private static OutputStreamWriter out = null;
    private static Path xmlFilePath = Daten.getMediathekXmlFilePath();

    public static synchronized void datenSchreiben() {
        xmlDatenSchreiben();
    }

    public static synchronized void exportPset(DatenPset[] pSet, String datei) {
        try {
            Log.systemMeldung("Pset exportieren");
            xmlFilePath = Paths.get(datei);
            xmlSchreibenStart();
            xmlSchreibenPset(pSet);
            xmlSchreibenEnde();
        } catch (Exception ex) {
            Log.fehlerMeldung(392846204, "IoXmlSchreiben.exportPset", ex, "nach: " + datei);
        }
    }

    private static void xmlDatenSchreiben() {
        try {
            Log.systemMeldung("Daten Schreiben");
            xmlSchreibenStart();

            writer.writeCharacters("\n\n");
            writer.writeComment("Abos");
            writer.writeCharacters("\n");
            xmlSchreibenAbo();

            writer.writeCharacters("\n\n");
            writer.writeComment("Blacklist");
            writer.writeCharacters("\n");
            xmlSchreibenBlackList();

            writer.writeCharacters("\n\n");
            writer.writeComment("Programmeinstellungen");
            writer.writeCharacters("\n");
            xmlSchreibenConfig(MVConfig.SYSTEM, Daten.mVConfig.getAll(), true);
            writer.writeCharacters("\n");

            writer.writeCharacters("\n\n");
            writer.writeComment("Programmsets");
            writer.writeCharacters("\n");
            xmlSchreibenProg();

            writer.writeCharacters("\n\n");
            writer.writeComment("Ersetzungstabelle");
            writer.writeCharacters("\n");
            xmlSchreibenErsetzungstabelle();

            writer.writeCharacters("\n\n");
            writer.writeComment("Downloads");
            writer.writeCharacters("\n");
            xmlSchreibenDownloads();

            writer.writeCharacters("\n\n");
            writer.writeComment("Update Filmliste");
            writer.writeCharacters("\n");
            xmlSchreibenFilmUpdateServer();

            writer.writeCharacters("\n\n");
            xmlSchreibenEnde();
        } catch (Exception ex) {
            Log.fehlerMeldung(656328109, "IoXml.xmlDatenSchreiben", ex);
        }
    }

    private static void xmlSchreibenStart() throws IOException, XMLStreamException {
        Log.systemMeldung("Start Schreiben nach: " + xmlFilePath.toAbsolutePath());
        out = new OutputStreamWriter(Files.newOutputStream(xmlFilePath), MSConst.KODIERUNG_UTF);

        XMLOutputFactory outFactory = XMLOutputFactory.newInstance();
        writer = outFactory.createXMLStreamWriter(out);
        writer.writeStartDocument(MSConst.KODIERUNG_UTF, "1.0");
        writer.writeCharacters("\n");//neue Zeile
        writer.writeStartElement(Konstanten.XML_START);
        writer.writeCharacters("\n");//neue Zeile
    }

    private static void xmlSchreibenErsetzungstabelle() {
        for (String[] sa : Daten.mVReplaceList.liste) {
            xmlSchreibenDaten(MVReplaceList.REPLACELIST, MVReplaceList.COLUMN_NAMES, sa, false);
        }
    }

    private static void xmlSchreibenProg() {
        //Proggruppen schreiben
        for (DatenPset datenPset : Daten.listePset) {
            xmlSchreibenDaten(DatenPset.PROGRAMMSET, DatenPset.COLUMN_NAMES_, datenPset.arr, false);
            for (DatenProg datenProg : datenPset.getListeProg()) {
                xmlSchreibenDaten(DatenProg.PROGRAMM, DatenProg.COLUMN_NAMES_, datenProg.arr, false);
            }
        }
    }

    private static void xmlSchreibenPset(DatenPset[] psetArray) {
        for (DatenPset pset : psetArray) {
            xmlSchreibenDaten(DatenPset.PROGRAMMSET, DatenPset.COLUMN_NAMES_, pset.arr, false);
            for (DatenProg datenProg : pset.getListeProg()) {
                xmlSchreibenDaten(DatenProg.PROGRAMM, DatenProg.COLUMN_NAMES_, datenProg.arr, false);
            }
        }
    }

    private static void xmlSchreibenDownloads() {
        //Abo schreiben
        for (DatenDownload download : Daten.listeDownloads) {
            if (download.isInterrupted()) {
                // unterbrochene werden gespeichert, dass die Info "Interrupt" erhalten bleibt
                xmlSchreibenDaten(DatenDownload.DOWNLOAD, DatenDownload.COLUMN_NAMES_, download.arr, false);
            } else if (!download.istAbo() && !download.isFinished()) {
                //Download, (Abo müssen neu angelegt werden)
                xmlSchreibenDaten(DatenDownload.DOWNLOAD, DatenDownload.COLUMN_NAMES_, download.arr, false);
            }
        }
    }

    private static void xmlSchreibenAbo() {
        //Abo schreiben
        for (DatenAbo datenAbo : Daten.listeAbo) {
            xmlSchreibenDaten(DatenAbo.ABO, DatenAbo.COLUMN_NAMES, datenAbo.arr, false);
        }
    }

    private static void xmlSchreibenBlackList() {
        //Blacklist schreiben
        for (DatenBlacklist blacklist : Daten.listeBlacklist) {
            xmlSchreibenDaten(DatenBlacklist.BLACKLIST, DatenBlacklist.BLACKLIST_COLUMN_NAMES, blacklist.arr, false);
        }
    }

    private static void xmlSchreibenFilmUpdateServer() throws XMLStreamException {
        //FilmUpdate schreiben
        writer.writeCharacters("\n");
        writer.writeComment("Akt-Filmliste");
        writer.writeCharacters("\n");
        for (DatenFilmlisteUrl datenUrlFilmliste : Daten.filmeLaden.getDownloadUrlsFilmlisten_akt()) {
            datenUrlFilmliste.arr[DatenFilmlisteUrl.FILM_UPDATE_SERVER_ART_NR] = DatenFilmlisteUrl.SERVER_ART_AKT;
            xmlSchreibenDaten(DatenFilmlisteUrl.FILM_UPDATE_SERVER, DatenFilmlisteUrl.FILM_UPDATE_SERVER_COLUMN_NAMES, datenUrlFilmliste.arr, false);
        }

        writer.writeCharacters("\n");
        writer.writeComment("Diff-Filmliste");
        writer.writeCharacters("\n");
        for (DatenFilmlisteUrl datenUrlFilmliste : Daten.filmeLaden.getDownloadUrlsFilmlisten_diff()) {
            datenUrlFilmliste.arr[DatenFilmlisteUrl.FILM_UPDATE_SERVER_ART_NR] = DatenFilmlisteUrl.SERVER_ART_DIFF;
            xmlSchreibenDaten(DatenFilmlisteUrl.FILM_UPDATE_SERVER, DatenFilmlisteUrl.FILM_UPDATE_SERVER_COLUMN_NAMES, datenUrlFilmliste.arr, false);
        }
    }

    private static void xmlSchreibenDaten(String xmlName, String[] xmlSpalten, String[] datenArray, boolean newLine) {
        final int xmlMax = datenArray.length;
        try {
            writer.writeStartElement(xmlName);
            if (newLine) {
                writer.writeCharacters("\n"); //neue Zeile
            }
            for (int i = 0; i < xmlMax; ++i) {
                if (!datenArray[i].equals("")) {
                    if (newLine) {
                        writer.writeCharacters("\t"); //Tab
                    }
                    writer.writeStartElement(xmlSpalten[i]);
                    writer.writeCharacters(datenArray[i]);
                    writer.writeEndElement();
                    if (newLine) {
                        writer.writeCharacters("\n"); //neue Zeile
                    }
                }
            }
            writer.writeEndElement();
            writer.writeCharacters("\n"); //neue Zeile
        } catch (Exception ex) {
            Log.fehlerMeldung(198325017, "IoXmlSchreiben.xmlSchreibenDaten", ex);
        }
    }

    private static void xmlSchreibenConfig(String xmlName, String[][] xmlSpalten, boolean newLine) {
        try {
            writer.writeStartElement(xmlName);
            if (newLine) {
                writer.writeCharacters("\n"); //neue Zeile
            }
            for (String[] xmlSpalte : xmlSpalten) {
                if (!xmlSpalte[1].equals("")) {
                    if (newLine) {
                        writer.writeCharacters("\t"); //Tab
                    }
                    writer.writeStartElement(xmlSpalte[0]);
                    writer.writeCharacters(xmlSpalte[1]);
                    writer.writeEndElement();
                    if (newLine) {
                        writer.writeCharacters("\n"); //neue Zeile
                    }
                }
            }
            writer.writeEndElement();
            writer.writeCharacters("\n"); //neue Zeile
        } catch (Exception ex) {
            Log.fehlerMeldung(951230478, "IoXmlSchreiben.xmlSchreibenConfig", ex);
        }
    }

    private static void xmlSchreibenEnde() throws Exception {
        writer.writeEndElement();
        writer.writeEndDocument();
        writer.flush();

        writer.close();
        out.close();

        Log.systemMeldung("geschrieben!");
    }
}

File: src/main/kotlin/mediathek/controller/IoXmlSchreiben.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package mediathek.controller

import mediathek.config.Daten
import mediathek.config.StandardLocations
import mediathek.daten.DatenProg
import mediathek.daten.DatenPset
import mediathek.tool.ReplaceList
import org.apache.logging.log4j.LogManager
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import javax.xml.stream.XMLOutputFactory
import javax.xml.stream.XMLStreamWriter

class IoXmlSchreiben(
    private val downloadStoragePath: Path = StandardLocations.getDownloadsFilePath(),
) {
    private val outFactory: XMLOutputFactory = XMLOutputFactory.newInstance()

    @Synchronized
    fun writeConfigurationFile(xmlFilePath: Path) {
        xmlDatenSchreiben(xmlFilePath)
        writeDownloads()
    }

    @Synchronized
    fun exportPset(pSet: Array<DatenPset>, datei: String) {
        val xmlFilePath = Paths.get(datei)
        try {
            Files.newOutputStream(xmlFilePath).use { output ->
                OutputStreamWriter(output, StandardCharsets.UTF_8).use { writer ->
                    val xmlWriter = outFactory.createXMLStreamWriter(writer)
                    logger.info("Pset exportieren nach: {}", xmlFilePath.toString())
                    logger.debug("Start Schreiben nach: {}", xmlFilePath.toAbsolutePath())

                    writeFileHeader(xmlWriter)

                    xmlSchreibenPset(xmlWriter, pSet)

                    writeFileEnd(xmlWriter)
                    logger.debug("geschrieben!")
                }
            }
        } catch (ex: Exception) {
            logger.error("nach {}", datei, ex)
        }
    }

    private fun writeFileHeader(writer: XMLStreamWriter) {
        writer.writeStartDocument(StandardCharsets.UTF_8.name(), "1.0")
        writeNewLine(writer)
        writer.writeStartElement("Mediathek")
        writeNewLine(writer)
    }

    private fun writeFileEnd(writer: XMLStreamWriter) {
        writer.writeEndElement()
        writer.writeEndDocument()
        writer.flush()
        writer.close()
    }

    private fun writeAbos(writer: XMLStreamWriter) {
        writer.writeCharacters("\n\n")
        writeNewLine(writer)

        for (datenAbo in Daten.getInstance().listeAbo) {
            datenAbo.writeToConfig(writer)
        }
    }

    private fun writeBlacklistRules(writer: XMLStreamWriter) {
        writer.writeCharacters("\n\n")
        writeNewLine(writer)

        // remove duplicates
        val distinctBlacklistRules = Daten.getInstance().listeBlacklist.stream().distinct().toList()
        for (rule in distinctBlacklistRules) {
            rule.writeToConfig(writer)
        }
    }

    private fun writeProgramSets(writer: XMLStreamWriter) {
        writer.writeCharacters("\n\n")
        writeNewLine(writer)
        // Proggruppen schreiben, bei Konfig-Datei
        for (datenPset in Daten.getInstance().listePset) {
            writeProgramSet(writer, datenPset)
            for (datenProg in datenPset.listeProg) {
                xmlSchreibenDaten(writer, DatenProg.TAG, DatenProg.XML_NAMES, datenProg.toArray(), true)
            }
        }
    }

    private fun writeReplacementTable(writer: XMLStreamWriter) {
        writer.writeCharacters("\n\n")
        // writer.writeComment("Ersetzungstabelle");
        writeNewLine(writer)

        for (values in ReplaceList.valuesForXml()) {
            xmlSchreibenDaten(writer, ReplaceList.REPLACELIST, ReplaceList.columnNames(), values, false)
        }
    }

    private fun writeDownloads() {
        try {
            DownloadStorage.write(downloadStoragePath, Daten.getInstance().listeDownloads)
        } catch (ex: Exception) {
            logger.error("writeDownloads error!", ex)
        }
    }

    private fun xmlSchreibenPset(writer: XMLStreamWriter, psetArray: Array<DatenPset>) {
        // wird beim Export Sets verwendet
        writer.writeCharacters("\n\n")
        for (pset in psetArray) {
            writeProgramSet(writer, pset)
            for (datenProg in pset.listeProg) {
                xmlSchreibenDaten(writer, DatenProg.TAG, DatenProg.XML_NAMES, datenProg.toArray(), true)
            }
            writer.writeCharacters("\n\n")
        }
    }

    private fun writeProgramSet(writer: XMLStreamWriter, pset: DatenPset) {
        try {
            val values = pset.toArray()
            writer.writeStartElement(DatenPset.TAG)
            writeNewLine(writer)

            for (i in values.indices) {
                writeProgramSetField(writer, DatenPset.XML_NAMES[i], values[i])
            }

            writer.writeEndElement()
            writeNewLine(writer)
        } catch (ex: Exception) {
            logger.error("writeProgramSet", ex)
        }
    }

    private fun writeProgramSetField(writer: XMLStreamWriter, xmlName: String, value: String?) {
        if (value.isNullOrEmpty()) {
            return
        }
        writer.writeCharacters("\t")
        writer.writeStartElement(xmlName)
        writer.writeCharacters(value)
        writer.writeEndElement()
        writeNewLine(writer)
    }

    private fun xmlSchreibenDaten(
        writer: XMLStreamWriter,
        xmlName: String,
        xmlColumns: Array<String>,
        data: Array<String>,
        newLine: Boolean,
    ) {
        try {
            writer.writeStartElement(xmlName)
            if (newLine) {
                writeNewLine(writer)
            }
            for (i in data.indices) {
                if (data[i].isNotEmpty()) {
                    if (newLine) {
                        writer.writeCharacters("\t") // Tab
                    }
                    writer.writeStartElement(xmlColumns[i])
                    writer.writeCharacters(data[i])
                    writer.writeEndElement()
                    if (newLine) {
                        writeNewLine(writer)
                    }
                }
            }
            writer.writeEndElement()
            writeNewLine(writer)
        } catch (ex: Exception) {
            logger.error("xmlSchreibenDaten", ex)
        }
    }

    private fun writeNewLine(writer: XMLStreamWriter) {
        writer.writeCharacters("\n") // neue Zeile
    }

    private fun xmlDatenSchreiben(xmlFilePath: Path) {
        logger.debug("Config Schreiben nach: {} startet", xmlFilePath.toAbsolutePath())

        try {
            Files.newOutputStream(xmlFilePath).use { output ->
                OutputStreamWriter(output, StandardCharsets.UTF_8).use { writer ->
                    val xmlWriter = outFactory.createXMLStreamWriter(writer)

                    writeFileHeader(xmlWriter)

                    writeAbos(xmlWriter)

                    writeBlacklistRules(xmlWriter)

                    writeProgramSets(xmlWriter)

                    writeReplacementTable(xmlWriter)

                    xmlWriter.writeCharacters("\n\n")

                    writeFileEnd(xmlWriter)
                }
            }
        } catch (ex: Exception) {
            logger.error("xmlDatenSchreiben", ex)
        }
    }

    companion object {
        private val logger = LogManager.getLogger(IoXmlSchreiben::class.java)
    }
}
